### DISCUSSION TITLE
Clip training - Diverging Eval Loss

### DISCUSSION THREAD

**Discussion by humanely** [huggingface/diffusers#7835] [Q&A]
I am training a CLIP model, but the eval_loss is diverging after 25K steps. Here are the few scenarios I have tried.
![image](https://github.com/huggingface/diffusers/assets/6862868/db5db81a-c447-46e9-b9cf-079dfb97b2da)
This was trained with following config:
CUDA_VISIBLE_DEVICES="0, 1, 2, 3"  accelerate launch  --mixed_precision="fp16"  run_clip.py \
    --max_grad_norm 0.9 \
    --num_train_epochs 1500 \
    --output_dir ~/clip-sa-base4 \
    --model_name_or_path /home/user/clip-sa2 \
    --tokenizer_name /home/user/clip-sanskrit-data/checkpoint-37920000 \
    --train_file /home/user/clip-sanskrit-data/annotations/train2017.json \
    --validation_file  /home/user/data/annotations/captions_val2017.json \
    --image_column image_path \
    --caption_column captions \
    --remove_unused_columns=False \
    --torch_compile=True \
    --load_best_model_at_end=True \
    --evaluation_strategy "steps" \
    --save_strategy  "steps" \
    --dataloader_drop_last True \
    --save_total_limit  10 \
    --no_cuda False \
    --do_train  --do_eval \
    --learning_rate="5e-5" --warmup_steps="0" --weight_decay 0.1 \
    --overwrite_output_dir

I added a 0.9 of max_grad_norm for second trial and weight decay of 0.2
![Pasted Graphic 1](https://github.com/huggingface/diffusers/assets/6862868/51c5382a-352b-40a4-b3a2-2e4e13b06630)

3rd Run: Change lr to 5e-7,  Weight decay to 0.1, max_grad_norm 1)
![Pasted Graphic 2](https://github.com/huggingface/diffusers/assets/6862868/5131d5ce-0c93-4890-a2b1-d7f0ee7d86d0)

Can somebody help me with how to reduce or eliminate this overfitting? TIA

**Comment by tolgacangoz** [comment]
Isn't [`run_clip.py` from `transformers`](https://huggingface.co/docs/transformers/main/en/model_doc/clip#resources:~:text=This%20example%20script%20shows%20how%20to%20train%20a%20CLIP%2Dlike%20vision%2Dtext%20dual%20encoder%20model%20using%20a%20pre%2Dtrained%20vision%20and%20text%20encoder%20using%20COCO%20dataset.)? It may be more appropriate to ask this question there, but I think we can also discuss about it.

**Comment by tolgacangoz** [comment]
In the second trial, increasing regularization, i.e., increasing weight decay and applying early stopping (especially this one) seems helped IMHO.

**Comment by humanely** [comment]
Yes, but excessive decay or early stop is not good approach.

**Reply by tolgacangoz** [reply]
I think `weight_decay=0.2` is not excessive here compared to `weight_decay=0.1`. Their plots seem similar. There is no underfitting.
I think just applying early stopping may be enough here. Why do you think that they are not good approaches?

**Comment by humanely** [comment]
I will setup wandb for this. I am new so will have to check how exactly to do it. I have tensorflow in it currently.

For weight_decay of 0.3, learning stops.
![image](https://github.com/huggingface/diffusers/assets/6862868/69a95409-2089-447d-b09d-7cd7e45fd6b7)

Early stopping might give you the best low evaluation scores but the search for global minima will remain equally hard. So early stopping is good for its worth. Do you suggest that do the early stop, and reuse the checkpoint for next set of trials?

**Reply by tolgacangoz** [reply]
[AFAIK, "getting stuck in a local minimum" doesn't seem to be much of an issue anymore.](https://www.quora.com/What-is-the-most-counterintuitive-result-in-deep-learning/answer/Yoshua-Bengio)
Could you evaluate the outputs of the CLIP obtained with early stopping?

**Reply by humanely** [reply]
Everytime after early stopping, it either stops after one earlystop without improvements for all variations of weight_decay.
Otherwise it keeps diverging.
![image](https://github.com/huggingface/diffusers/assets/6862868/e2befe83-195e-4eaf-82d1-ab7ab724bdda)

**Reply by tolgacangoz** [reply]
Early stopping is the final point. One shouldn't train further.
Could you evaluate the outputs of the CLIP obtained with early stopping manually? Please give it some input and evaluate.

**Reply by humanely** [reply]
I tried evaluating. But this became another issue. Not able to load the tokenizer.

[https://discuss.huggingface.co/t/clip-the-backend-tokenizer-provided-does-not-match-the-expected-format/85718](https://discuss.huggingface.co/t/clip-the-backend-tokenizer-provided-does-not-match-the-expected-format/85718)
